Logging feature issue #246

namsaraeva opened the pull request:
**Description of changes**

Created a new logging instance for future ```logg``` messages. 

**Technical details**

Heavily adapted from Cellrank and Scanpy. 

**Additional context**

Will add logg.info() messages in every method where it is needed in the next days/weeks.

Zethson commented on ehrapy/logger.py:
```diff
@@ -0,0 +1,199 @@
+        return result
+
+
+_DEPENDENCIES_NUMERICS = [
```
We should either get rid of this here or in the meta_information.py script -> https://github.com/theislab/ehrapy/blob/development/ehrapy/core/meta_information.py

I think that for you now it is easier if you delete the duplications in the logger.py script and add any missing things (e.g. `_DEPENDENCIES_PLOTTING` etc to the existing meta_information.py script.

Zethson commented on ehrapy/logger.py:
```diff
@@ -0,0 +1,199 @@
+    )
+
+
+def print_version_and_date():
```
This is now also a duplicate and should get removed.

Zethson reviewed (changes requested):
In general, please remove all duplicates that you introduced with https://github.com/theislab/ehrapy/blob/development/ehrapy/core/meta_information.py

Besides this you are on track :) Thank you!

Zethson commented on ehrapy/logger.py:
```diff
@@ -0,0 +1,156 @@
+        return result
+
+
+def _versions_dependencies(dependencies):
```
Could replace the import here? https://github.com/theislab/ehrapy/blob/b4bfdf6e68201c97afb94fa17a283c6ac679102c/ehrapy/core/meta_information.py#L11

Not important though likely....

namsaraeva replied:
Do you want to import `_versions_dependencies ` from ehrapy now?

Zethson replied:
If they are equivalent it might be nicer, yeah!

Zethson commented on ehrapy/logger.py:
```diff
@@ -0,0 +1,156 @@
+    deep: Optional[str] = None,
+    extra: Optional[dict] = None,
+) -> datetime:
+    """
```
Please use the Google Docstring format :)

Zethson reviewed (commented):
Good! Now I guess you need to create the actual logging messages :)

Zethson commented on ehrapy/logger.py:
```diff
@@ -104,24 +104,16 @@ def error(
+    Args:
+        msg: Message to display.
+        time: A time in the past. If this is passed, the time difference from then to now is appended to `msg` as ` (HH:MM:SS)`.
+         If `msg` contains `{time_passed}`, the time difference is instead inserted at that position
```
Usually these need to be aligned

->

```
time: BLABLABLA
         CONTINUATION OF BALBLA
```

Else the might not render properly

Zethson commented on ehrapy/data/_dataloader.py:
```diff
@@ -61,3 +61,7 @@ def download(
         output_path = output_path or tempfile.gettempdir()
         shutil.unpack_archive(download_to_path, output_path)
+
+    logg.info(
```
Don't think that we need logging messages for this or do we? @Imipenem opinion?

Imipenem replied:
Hm, could actually be helpful (especially since this code part was a bit tricky initially), so I'd say it's fine.

Imipenem commented on ehrapy/logger.py:
```diff
@@ -0,0 +1,149 @@
+        time: datetime = None,
+        deep: Optional[str] = None,
+    ) -> datetime:
+        from ehrapy import settings
```
We are using this in multiple places across the whole file; maybe we should just import this one once at the top?

namsaraeva replied:
Yes, makes sense, will be done.

Zethson commented on ehrapy/logger.py:
```diff
@@ -0,0 +1,149 @@
+from typing import Optional
+
+HINT = (INFO + DEBUG) // 2
+logging.addLevelName(HINT, "HINT")
```
I'm not sure either. It's stolen from cellrank :)

Zethson commented on ehrapy/anndata/anndata_ext.py:
```diff
@@ -79,6 +80,12 @@ def df_to_anndata(
     )
 
+    logg.info(
+        f"Transformed given dataframe into an AnnData object with n_obs x n_vars = `{adata.n_obs}` x `{adata.n_vars}`"
```
```suggestion
        f"Transformed passed dataframe into an AnnData object with n_obs x n_vars = `{adata.n_obs}` x `{adata.n_vars}`"
```

Zethson commented on ehrapy/anndata/anndata_ext.py:
```diff
@@ -97,13 +104,19 @@ def _move_columns_to_obs(df: pd.DataFrame, columns_obs_only: list[str] | None) -
             obs = obs.set_index(df.index.map(str))
             df = df.drop(columns_obs_only, axis=1)
+            logg.info(
+                f"Columns `{columns_obs_only}` were successfully moved to `obs`"
```
```suggestion
                f"Added columns `{columns_obs_only}` to `obs`"
```

Zethson commented on ehrapy/anndata/anndata_ext.py:
```diff
@@ -97,13 +104,19 @@ def _move_columns_to_obs(df: pd.DataFrame, columns_obs_only: list[str] | None) -
     else:
         obs = pd.DataFrame(index=df.index.map(str))
+        logg.info(
+            f"All columns were successfully moved to `obs`"
```
```suggestion
            f"Added all columns to `obs`"
```

Zethson commented on ehrapy/anndata/anndata_ext.py:
```diff
@@ -151,18 +168,24 @@ def anndata_to_df(
+            f"Added `{var_cols}` columns to `X`."
+        )
+
+    logg.info(
```
I'd remove this logging message.

Zethson commented on ehrapy/anndata/anndata_ext.py:
```diff
@@ -188,17 +212,19 @@ def move_to_obs(adata: AnnData, to_obs: list[str] | str, copy: bool = False) ->
     adata.uns["non_numerical_columns"] = updated_non_num_uns
 
+    logg.info(
+        f"Moved `{to_obs}` columns to `obs`."
```
```suggestion
        f"Added `{to_obs}` to `obs`."
```

Zethson commented on ehrapy/anndata/anndata_ext.py:
```diff
@@ -212,6 +238,10 @@ def move_to_x(adata: AnnData, to_x: list[str] | str) -> AnnData:
     new_adata.uns["non_numerical_columns"] = adata.uns["non_numerical_columns"] + non_num_columns_moved
 
+    logg.info(
+        f"Moved `{to_x}` features to `X`."
```
```suggestion
        f"Added `{to_x}` features to `X`."
```

Zethson reviewed (changes requested):
I'll be a bit lazy and not suggest the same thing everywhere.
This looks good. Basically, my request is that we try to use the same language when things have been added somewhere. Consistency.

Instead of "move" or whatever I'd always use "Add X to Y".

namsaraeva commented:
About the `rich` module: it is not that trivial as I thought, so I reverted the changes that I introduced by replacing `print()` with `logg.warning()`. I'll open a new issue for this: introduce coloured messages into our logging instance.

namsaraeva commented:
Do the messages appear red? I modified the LogFormatter and added some colours for different message types but it does not seem to work on my laptop. Also there is a safety issue: `Vulnerability found in ipywidgets version 7.6.6`. Any ideas?
